application/apps/indexer/sources/src/serial/serialport.rs

```diff
@@ -0,0 +1,131 @@
+use crate::ByteSource;
+use crate::{Error as SourceError, ReloadInfo, SourceFilter};
+use async_trait::async_trait;
+use buf_redux::Buffer;
+use std::time::Duration;
+use tokio::io::AsyncReadExt;
+use tokio_serial::{SerialPortBuilderExt, SerialStream};
+
+const TIMEOUT: u64 = 100;
+
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum SerialError {
+    #[error("Serial setup problem: {0}")]
+    Setup(String),
+    #[error("Unrecoverable serial error: {0}")]
+    Unrecoverable(String),
+}
+
+pub struct SerialSource {
+    port: SerialStream,
+    buffer: Buffer,
+    read: [u8; 1],
+    amount: usize,
+}
+
+impl SerialSource {
+    pub fn new(path: &str, baud_rate: u32) -> Result<Self, SerialError> {
+        Ok(Self {
+            buffer: Buffer::new(),
+            read: [0; 1],
+            amount: 0,
+            port: tokio_serial::new(path, baud_rate)
+                .open_native_async()
+                .map_err(|e| SerialError::Unrecoverable(format!("Could not open port: {}", e)))?,
+        })
+    }
+}
+
+#[async_trait]
+impl ByteSource for SerialSource {
+    async fn reload(
+        &mut self,
+        _filter: Option<&SourceFilter>,
+    ) -> Result<Option<ReloadInfo>, SourceError> {
+        let mut copied: usize = 0;
+        let mut buffer_error: String = "".to_string();
+        let timeout = tokio::time::timeout(Duration::from_millis(TIMEOUT), async {
+            loop {
+                if let Err(err) = self.port.read_exact(&mut self.read).await {
+                    buffer_error = format!("read from serial stream failed: {}", err);
+                };
+                copied += self.buffer.copy_from_slice(&self.read);
+            }
```

Comment:
wait a second...is this really working? we basically never break out of the loop until we timeout. But in that case we always return `ReloadInfo::new(0, 0, 0, None)`